core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,288 @@
+
+    def restrict(restricted: Option[CacheHint]): CachePolicy = restricted match {
+      case Some(h) => copy(hint = mostRestrictive(hint, h))
+      case None    => this
+    }
+
+    private def mostRestrictive(
+      parent: CacheHint,
+      current: CacheHint
+    ): CacheHint = {
+      val scope         = List(parent.scope, current.scope).flatten.minOption
+      val shouldInherit = current.inheritMaxAge
+      val maxAge        =
+        if (shouldInherit) parent.maxAge
+        else
+          (current.maxAge, parent.maxAge) match {
+            case (Some(a), Some(b)) => Some(a min b)
+            case (Some(a), None)    => Some(a)
+            case (None, Some(b))    => Some(b)
+            case (None, None)       => None
+          }
+      CacheHint(maxAge, scope, shouldInherit)
+    }
+  }
+
+  case class CacheHint(
+    maxAge: Option[Duration] = None,
+    scope: Option[CacheScope] = None,
+    inheritMaxAge: Boolean = false
+  )
+
+  object CacheHint {
+    val default: CacheHint = CacheHint()
+  }
+
+  sealed trait CacheScope
+  object CacheScope {
+
+    @GQLName("PRIVATE")
+    case object Private extends CacheScope {
+      override def toString: String = "PRIVATE"
+    }
+
+    @GQLName("PUBLIC")
+    case object Public extends CacheScope {
+      override def toString: String = "PUBLIC"
+    }
+
+    implicit val ord: Ord[CacheScope] = Ord.make {
+      case (Private, Private) => Ordering.Equals
+      case (Public, Public)   => Ordering.Equals
+      case (Private, Public)  => Ordering.LessThan
+      case (Public, Private)  => Ordering.GreaterThan
+    }
+
+    implicit val ordering: scala.math.Ordering[CacheScope] = ord.toScala
+
+    implicit val schema: Schema[Any, CacheScope] = Schema.gen
+  }
+
+  private case class Default[-R1](settings: CacheSettings) extends GraphQLAspect[Nothing, R1] {
+    private val _typeCache = new ConcurrentHashMap[String, Option[CacheHint]]()
+
+    def apply[R <: R1](gql: GraphQL[R]): GraphQL[R] = {
+      val wrapper = EffectfulWrapper(
+        Ref
+          .make(CachePolicy(CacheHint.default))
+          .map(state => staticWrapper(state) |+| fieldWrapper(state) |+| overallWrapper(state))
+      )
+
+      gql
+        .withAdditionalDirectives(cacheDirectives)
+        .withAdditionalTypes(CacheScope.schema.toType_() :: Nil)
+        .withWrapper(wrapper)
+    }
+
+    private[caliban] def cacheHintFromType(typ: __Type): Option[CacheHint] = {
+      val key = typ.name.getOrElse(typ.toString)
+      Option(_typeCache.get(key)) match {
+        case Some(Some(hint)) => Some(hint)
+        case Some(None)       => None
+        case None             =>
+          typ.directives match {
+            case Some(dirs) =>
+              val hint = extractCacheDirective(dirs)
+              hint.foreach(h => _typeCache.put(key, Some(h)))
+              hint
+            case None       =>
+              _typeCache.put(key, None)
+              None
+          }
+      }
+    }
```

Comment:
Assuming that we don't need the `.foreach` part as per my comment above, perhaps consider using `computeIfAbsent` instead:

```suggestion
    private[caliban] def cacheHintFromType(typ: __Type): Option[CacheHint] = {
      val key = typ.name.getOrElse(typ.toString)
      _typeCache.computeIfAbsent(key, _ => typ.directives.flatMap(extractCacheDirective))
    }
```